scylla-cdc-printer/src/main.rs

```diff
@@ -1,11 +1,73 @@
 use std::error::Error;
-use std::time::Duration;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use chrono::Duration;
+use core::time;
+use scylla::{Session, SessionBuilder};
+use tokio::time::sleep;
+
+use scylla_cdc::reader::Reader;
+use scylla_cdc::stream_generation::GenerationFetcher;
+
+const SLEEP_INTERVAL: u64 = 1;
+const WINDOW_SIZE: i64 = 2;
+const SAFETY_INTERVAL: i64 = 1;
+
+struct CDCLogPrinter {
+    session: Arc<Session>,
+    sleep_interval: time::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+}
+
+impl CDCLogPrinter {
+    fn new(session: &Arc<Session>) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            sleep_interval: time::Duration::new(SLEEP_INTERVAL, 0),
+            window_size: chrono::Duration::seconds(WINDOW_SIZE),
+            safety_interval: chrono::Duration::seconds(SAFETY_INTERVAL),
+        }
+    }
+
+    async fn run(&self, keyspace: String, table_name: String, start_timestamp: chrono::Duration) -> Result<(), Box<dyn Error>> {
+        let fetcher = GenerationFetcher::new(&self.session);
+        let mut generation = fetcher.fetch_generation_by_timestamp(&start_timestamp).await?.unwrap_or(loop {
+            match fetcher.fetch_all_generations().await?.last() {
+                None => { sleep(self.sleep_interval).await; }
+                Some(generation) => break generation.clone()
+            }
+        });
```

Comment:
The expression passed to `unwrap_or` will always be evaluated. This means that `fetch_all_generations` will always be called, sometimes unnecessarily.

I suggest moving the logic responsible for getting the first generation to a helper function.